### PR TITLE
DAOS-1386 test container snapshot info DAOS-1402 DAOS-1390

### PR DESCRIPTION
**Pull request by dinghwah**
Updated file: src/utils/py/daos_api.py
New     file: src/tests/ftest/container/snapshot_info_test.py
New     file: src/tests/ftest/container/snapshot_info_test.yaml

Signed-off-by: Ding Ho <ding-hwa.ho@intel.com>

### CODE REVIEW

File: src/tests/ftest/container/snapshot_info_test.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+
+
+    def test_snapshot_info(self):
+        """
+        Test ID: DAOS-1386 Test container SnapShot information
+                 DAOS-1371 Test list snapshots
+                 DAOS-1395 Test snapshot destroy
+                 DAOS-1402 Test creating multiple snapshots
+        Test Description:
+                (1)Create an object, write random adata into it, and take
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.py]: typo "adata"

File: src/tests/ftest/container/snapshot_info_test.py [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+                   does a commit when the update is complete.
+                (3)Verify the data in the snapshot is the original data.
+                   Get a handle for the snapshot and read the object at dkey,
+                   akey. Compare it to the originally written data.
+                (4)List the snapshot and make sure it reflects the original
+                   epoch.
+                   ==>Repeat step(1) to step(4) for multiple snapshot tests.
+                (5)Verify the snapshots data.
+                (6)Destroy the snapshot.
+                (7)Check if still able to Open the destroied snapshot and
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.py]: typo "destroied"

File: src/tests/ftest/container/snapshot_info_test.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+                   Verify the snapshot removed from the snapshot list.
+        Use Cases:
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.py]: same as before.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+        #(6)Destroy the snapshot
+        self.log.info("==(6)Destroy snapshot  epoch: %s", epoch)
+        try:
+            snapshot.destroy(self.container.coh, epoch)
+            self.log.info("==Snapshot successfully destroyed")
+        except Exception as error:
+            self.fail(
+                "##Error on snapshot.destroy: %s", str(error))
+
+        #(7)Check if still able to Open the destroied snapshot and
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.py]: typo "destroied"

File: src/tests/ftest/container/snapshot_info_test.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# change host names to your reserved nodes, the
+# required quantity is indicated by the placeholders
+hosts:
+    test_machines1:
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.yaml]: Can be removed. Don't need extra level in the tree for nodes.

File: src/tests/ftest/container/snapshot_info_test.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      test_machines:
+        - boro-A
+        - boro-B
+timeout: 100
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.yaml]: Is this timeout atleast twice the actual runtime required for each test?

File: src/tests/ftest/container/snapshot_info_test.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+faults: !mux
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.yaml]: Don't need mux here

File: src/tests/ftest/container/snapshot_info_test.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    no_faults:
+      fault_list: []
+server_config:
+   name: daos_server
+poolparams:
+   createmode:
+     mode: 511
+   createset:
+     setname: daos_server
+   createsize: !mux
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.yaml]: don't need mux here

File: src/tests/ftest/container/snapshot_info_test.py [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+        self.log.info("==(5)Snapshot with a NULL epoch.")
+        if self.invalid_snapshot_test(self.container.coh, None):
+            self.log.info("==>Negative test 3, expecting failed on taking "
+                          "snapshot with a NULL epoch.")
+        else:
+            self.fail("##(5)Negative test 3 passing, expecting failed on "
+                      "taking snapshot with a NULL epoch.")
+
+
+    def test_snapshot_info(self):
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_info_test.py]: Seems like this test is doing more than just testing snapshot information functionality. Hence the test name can be a bit misleading. We can change it to something more meaningful here. Like - snapshot_functionality or something. Also same goes for the name of the test and yaml files. The tests in side these files are doing more than just "info" test. So having something more meaning will be nice.
Can be "snapshot_functionality_test.py" If this covers all the important functional tests for snapshot. If you think some better please go ahead with that.

**Comment by dinghwah** [comment]
Saurabh, 
Thanks for the review and comments, I have addressed all the comments on the updated test scripts. I also change the test script name to:   snapshot_functionality_test.py and .yaml.
Please review.
Thanks, Ding.

**Comment by dinghwah** [comment]
Saurabh, 
Thanks for the review and comments, I have addressed all the comments on the updated test scripts. I also change the test script name to:   snapshot_functionality_test.py and .yaml.
Please review.
Thanks, Ding.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+            except Exception as error:
+                self.fail(
+                    "##Test was unable to list the snapshot: {}"
+                    .format(str(error)))
+
+        #(5)Verify the snapshots data
+        try:
+            for ind in range(0, len(self.container_epoch_list)):
+                epoch = self.container_epoch_list[ind]
+                ss = self.snapshot_list[ind]
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+  portions thereof marked with this legend must also reproduce the markings.
+"""
+from __future__ import print_function
+import os
+import time
+import traceback
+import sys
+import random
+import string
+import json
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+            container.destroy()
+        except DaosApiError as excep:
+            self.log.info(excep)
+            self.log.info(traceback.format_exc())
+            self.fail("##Snapshot info test failed on cleanUp.")
+        finally:
+            # cleanup the pool
+            pool.disconnect()
+            pool.destroy(1)
+
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+        self.log.info("self.container.info.es_lhe=        %s"
+            ,self.container.info.es_lhe)
+        self.log.info("self.container.info.es_ghce=       %s"
+            ,self.container.info.es_ghce)
+        self.log.info("self.container.info.es_glre=       %s"
+            ,self.container.info.es_glre)
+        self.log.info("self.container.info.es_ghpce=      %s"
+            ,self.container.info.es_ghpce)
+        self.log.info("self.container.info.ci_nsnapshots= %s"
+            ,self.container.info.ci_nsnapshots)
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+        self.log.info("self.container.info.ci_snapshots=  %s"
+            ,self.container.info.ci_snapshots)
+        self.log.info("self.container.info.ci_min_slipped_epoch= %s"
+            ,self.container.info.ci_min_slipped_epoch)
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 15]
```diff
@@ -0,0 +1,522 @@
...
+          testcases.
+    :avocado: recursive
+    """
+
+    def setUp(self):
+        """
+        set up method
+        """
+        super(Snapshot, self).setUp()
+        # get paths from the build_vars generated by build
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Thanks for the review and comments,
Please see the new commite (#6) with a new files snapshot_functionality_test.py and .yaml

[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 16]
```diff
@@ -0,0 +1,522 @@
...
+                      , self.container.info.ci_nsnapshots)
+        self.log.info("self.container.info.ci_snapshots=  %s"
+                      , self.container.info.ci_snapshots)
+        self.log.info("self.container.info.ci_min_slipped_epoch= %s"
+                      , self.container.info.ci_min_slipped_epoch)
+        self.log.info("==================================")
+
+    def display_snapshot(self, snapshot):
+        """
+        To display the snapshot information.
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py [hunk 17]
```diff
@@ -0,0 +1,589 @@
...
+            self.log.info(excep)
+            self.log.info(traceback.format_exc())
+            self.fail("##Snapshot negative test failed on cleanUp.")
+        finally:
+            # cleanup the pool
+            pool.disconnect()
+            pool.destroy(1)
+
+
+    def test_snapshot_info(self):
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py (hunk 1 above)
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Thanks. Upadted.

File: src/tests/ftest/container/snapshot_info_test.py (hunk 3 above)
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_info_test.py (hunk 4 above)
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_info_test.py]: Updated.

File: src/tests/ftest/container/snapshot_functionality_test.yaml [hunk 18]
```diff
@@ -0,0 +1,31 @@
+# change host names to your reserved nodes, the
+# required quantity is indicated by the placeholders
+hosts:
+    test_machines:
+        - boro-A
+        - boro-B
+timeout: 100
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_functionality_test.yaml]: Can you please suggest if this timeout is at least twice the total run time of this test?

File: src/tests/ftest/container/snapshot_functionality_test.yaml [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+faults: !mux
```
Comment (saurabhtandan) [inline src/tests/ftest/container/snapshot_functionality_test.yaml]: Don't think you need mux here. Can you please explain if you have it for any reason?

Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.yaml]: The run-time is about 250-seconds.  Have updated with 500.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 20]
```diff
@@ -0,0 +1,529 @@
...
+
+
+    def test_snapshot_info(self):
+        """
+        Test ID: DAOS-1386 Test container SnapShot information
+                 DAOS-1371 Test list snapshots
+                 DAOS-1395 Test snapshot destroy
+                 DAOS-1402 Test creating multiple snapshots
+        Test Description:
+                (1)Create an object, write random adata into it, and take
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Thanks.

File: src/tests/ftest/server/metadata.yaml [hunk 21]
```diff
@@ -3,10 +3,10 @@
 
 hosts:
   servers:
-    - boro-A
-    - boro-B
+    - boro-34
```
Comment (saurabhtandan) [inline src/tests/ftest/server/metadata.yaml]: Please undo the changes and resubmit

File: src/tests/ftest/soak/soak.yaml [hunk 22]
```diff
@@ -1,7 +1,7 @@
 hosts:
    test_machines:
         daos_servers:
-            - boro-A
+            - boro-65
```
Comment (saurabhtandan) [inline src/tests/ftest/soak/soak.yaml]: same as above

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 23]
```diff
@@ -0,0 +1,532 @@
...
+            time.sleep(2)
+            self.container.destroy()
+        except DaosApiError as excep:
+            self.log.info(excep)
+            self.log.info(traceback.format_exc())
+            self.fail("##Snapshot test failed on cleanUp.")
+        finally:
+            # cleanup the pool
+            self.pool.disconnect()
+            self.pool.destroy(1)
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Thanks for the comment, please see the new commit which has the fixes.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 24]
```diff
@@ -0,0 +1,539 @@
...
+                 DAOS-1395 Test snapshot destroy
+                 DAOS-1402 Test creating multiple snapshots
+        Test Description:
+                (1)Create an object, write random data into it, and take
+                   a snapshot.
+                (2)Make changes to the data object. The write_an_obj function
+                   does a commit when the update is complete.
+                (3)Verify the data in the snapshot is the original data.
+                   Get a handle for the snapshot and read the object at dkey,
+                   akey. Compare it to the originally written data.
```
Comment (johannlombardi) [inline src/tests/ftest/container/snapshot_functionality_test.py]: Good. You should just merge this with the previous test and do:
1. write data
2. take snapshot
3. read data back & verify content
4. overwrite data
5. read data from snapshot & verify content
...

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 25]
```diff
@@ -0,0 +1,539 @@
...
+        snapshot_loop = 10
+        obj_cls = self.params.get("obj_class", '/run/object_class/*')
+        akey = self.params.get("akey", '/run/snapshot/*', default="akey")
+        dkey = self.params.get("dkey", '/run/snapshot/*', default="dkey")
+        rand_str = lambda n: ''.join([random.choice(string.lowercase)
+                                      for i in xrange(n)])
+        #
+        #Test loop for creat, modify and snapshot object in the DAOS container.
+        #
+        while snapshot_index <= snapshot_loop:
```
Comment (johannlombardi) [inline src/tests/ftest/container/snapshot_functionality_test.py]: sorry, i won't understand what this snapshot loop is about. Could you please elaborate?

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 26]
```diff
@@ -0,0 +1,539 @@
...
+            #Save snapshot test data
+            coh_list.append(self.container.coh)
+            container_epoch_list.append(epoch)
+            snapshot_list.append(snapshot)
+            test_data.append(thedata)
+            snapshot_index += 1
+
+            #(2)Make changes to the data object. The write_an_obj function does
+            #   a commit when the update is complete
+            more_transactions = 100
```
Comment (johannlombardi) [inline src/tests/ftest/container/snapshot_functionality_test.py]: sorry, don't get what those transactions are about.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 27]
```diff
@@ -0,0 +1,539 @@
...
+            self.log.info("     snapshot.epoch=%s"
+                          , snapshot.epoch)
+            ##self.log.info("tickets already assigned DAOS-2390 DAOS-2392")
+            #if snapshot.epoch != reported_epoch:
+            #    raise Exception("##The snapshot epoch returned from "
+            #        "snapshot list is not the same as the original"
+            #        "epoch list is not the same as the original epoch"
+            #        "snapshotted.")
+            self.log.info("==After 10 additional commits the snapshot is "
+                          "still available")
```
Comment (johannlombardi) [inline src/tests/ftest/container/snapshot_functionality_test.py]: Not sure what this is about.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 28]
```diff
@@ -0,0 +1,539 @@
+#!/usr/bin/python
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Updated.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 29]
```diff
@@ -0,0 +1,539 @@
...
+
+    Test Class Description:
+          Start DAOS servers, set up the pool and container for the above
+          snapshot Epic and Testcases, including snapshot basic, container
+          information, list, creation and destroy.
+    :avocado: recursive
+    """
+    DAOS_PC_RO = int(1 << 0)
+    DAOS_PC_RW = int(1 << 1)
+    DAOS_PC_EX = int(1 << 2)
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Updated

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 30]
```diff
@@ -0,0 +1,539 @@
...
+    def tearDown(self):
+        """
+        tear down method
+        """
+        try:
+            if self.container:
+                self.container.close()
+
+            # wait a few seconds and then destroy
+            time.sleep(5)
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Removed the unneeded sleep.
Thanks,

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 31]
```diff
@@ -0,0 +1,539 @@
...
+                      , snapshot.context.ftable["list-attr"])
+        self.log.info("snapshot.context.ftable[test-event]=%s"
+                      , snapshot.context.ftable["test-event"])
+        self.log.info("snapshot.name=  %s", snapshot.name)
+        self.log.info("snapshot.epoch= %s", snapshot.epoch)
+        self.log.info("==================================")
+
+    def take_snapshot(self, container, epoch):
+        """
+        To take a snapshot on the container with current epoch.
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: snapshot took container-handle and container-epoch as input parameters.
and snapshot.epoch is an output parameter.

File: src/tests/ftest/container/snapshot_functionality_test.py [hunk 32]
```diff
@@ -0,0 +1,539 @@
...
+                      "taking snapshot with a NULL container.coh.")
+
+        #(5)Test snapshot with an invalid epoch
+        self.log.info("==(5)Snapshot with a NULL epoch.")
+        if self.invalid_snapshot_test(self.container.coh, None):
+            self.log.info("==>Negative test 3, expecting failed on taking "
+                          "snapshot with a NULL epoch.")
+        else:
+            self.fail("##(5)Negative test 3 passing, expecting failed on "
+                      "taking snapshot with a NULL epoch.")
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Snapshot is failing on a NULL epoch (as input parameter).

File: src/tests/ftest/container/snapshot_functionality_test.py (hunk 24 above)
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: Updated.

File: src/tests/ftest/container/snapshot_functionality_test.py (hunk 26 above)
Comment (dinghwah) [reply src/tests/ftest/container/snapshot_functionality_test.py]: As discussed in the review meeting.

File: src/utils/py/daos_api.py [hunk 33]
```diff
@@ -1876,13 +1876,15 @@ def create(self, coh, epoch):
     #  snapshots and the epochs and names lists. See description of
     #  daos_cont_list_snap in src/include/daos_api.h. This must be done for
     #  DAOS-1336 Verify container snapshot info.
-    def list(self, coh):
+    def list(self, coh, epoch=None):
```
Comment (johannlombardi) [inline src/utils/py/daos_api.py]: From the C API:
  * \param[out]  epochs  preallocated array of epochs to store snapshots.

So this is actually where we return the snapshot list.

File: src/utils/py/daos_api.py [hunk 34]
```diff
@@ -1876,13 +1876,15 @@ def create(self, coh, epoch):
...
         list.
         coh --ctype.u_long handle on an open container
         Returns the value of the epoch for this DaosSnapshot object.
         """
         func = self.context.get_function('list-snap')
+        if epoch is None:
+            epoch = self.epoch
         num = ctypes.c_uint64(1)
         epoch = ctypes.c_uint64(self.epoch)
         anchor = Anchor()
```
Comment (johannlombardi) [inline src/utils/py/daos_api.py]: I think we should just report back a dictionary with the list of epochs that have been snapshotted.

File: src/utils/py/daos_api.py [hunk 35]
```diff
@@ -1893,13 +1895,15 @@ def list(self, coh):
                                .format(retcode))
         return epoch.value
 
-    def open(self, coh):
+    def open(self, coh, epoch=None):
```
Comment (johannlombardi) [inline src/utils/py/daos_api.py]: If the intent is really to open a snapshot, then the epoch should be mandatory there.

File: src/utils/py/daos_api.py [hunk 36]
```diff
@@ -1908,7 +1912,7 @@ def open(self, coh):
                                .format(retcode))
         return txhndl
 
-    def destroy(self, coh, evnt=None):
+    def destroy(self, coh, epoch=None, evnt=None):
```
Comment (johannlombardi) [inline src/utils/py/daos_api.py]: Same comment here.

File: src/utils/py/daos_api.py [hunk 37]
```diff
@@ -1917,6 +1921,8 @@ def destroy(self, coh, evnt=None):
         # need container handle coh, and the epoch range
         """
         func = self.context.get_function('destroy-snap')
+        if epoch is None:
+            epoch = self.epoch
```
Comment (johannlombardi) [inline src/utils/py/daos_api.py]: We should remove the notion of self.epoch totally IMO.
